src/store/writer.rs

```diff
@@ -21,30 +24,67 @@ use crate::DocId;
 pub struct StoreWriter {
     compressor: Compressor,
     block_size: usize,
-    doc: DocId,
-    first_doc_in_block: DocId,
-    offset_index_writer: SkipIndexBuilder,
-    writer: CountingWriter<WritePtr>,
+    num_docs_in_current_block: DocId,
     intermediary_buffer: Vec<u8>,
     current_block: Vec<u8>,
+
+    writer: WritePtr,
+    // the channel to communicate with the compressor thread.
+    compressor_sender: SyncSender<BlockCompressorMessage>,
+    // the channel to receive data to write from the compressor thread.
+    data_receiver: Receiver<OwnedBytes>,
+
+    // the handle to check for errors on the thread
+    compressor_thread_handle: JoinHandle<io::Result<(DocStoreFooter, SkipIndexBuilder)>>,
+}
+
+enum BlockCompressorMessage {
+    AddBlock(DocumentBlock),
+    Stack((StoreReader, DocumentBlock)),
 }
 
 impl StoreWriter {
     /// Create a store writer.
     ///
     /// The store writer will writes blocks on disc as
     /// document are added.
-    pub fn new(writer: WritePtr, compressor: Compressor, block_size: usize) -> StoreWriter {
-        StoreWriter {
+    pub fn new(
+        writer: WritePtr,
+        compressor: Compressor,
+        block_size: usize,
+    ) -> io::Result<StoreWriter> {
+        let thread_builder = thread::Builder::new().name("docstore compressor thread".to_string());
+
+        // Data channel to send fs writes, to write only from current thread
```

Comment:
interesting... why do we want to write only from current thread?

Reply:
Even though in tantivy we create a separate file, which is fine for a separate thread to write into, the `Directory` trait itself doesn't require explicitly thread safe writes. `TerminatingWrite` is not Send, so the current contract is that the writers stay on the same thread.

It's unlikely to be an issue, but it could be two threads write next to each other at the same page or cache line. So depending on the write target and buffer synchronization between the threads, it could be that they overwrite each others data. Since rust doesn't cover race conditions outside it's memory model, e.g. Files, I'm extra careful there.

Reply:
That makes sense. Can we make `TerminatingWrite: Send` and simplify the code though?

Reply:
Ah so you just offload compression to a different thread. Writing is still done in the same place. Maybe it is clearer that way let me keep on reading on.